File: javascript/daydream.js
function main(input) {
  const reverse = (s) => {
    return s.split('').reverse().join('')
  }

  const WORDS = ['dream', 'dreamer', 'erase', 'eraser'].map(reverse)
  const S = reverse(input.split('\n')[0])

  var index = 0

  while (true) {
    var w = WORDS.find((w) => S.startsWith(w, index))
    if (!w) break
    index += w.length
  }

  if (S.length === index) {
    console.log('YES')
  } else {
    console.log('NO')
  }
}
main(require('fs').readFileSync('/dev/stdin', 'utf8'))

File: go/daydream.go
package main

import (
	"bytes"
	"fmt"
)

func main() {
	var S string
	fmt.Scan(&S)
	sb := []byte(S)
	rev(sb)

	w := [][]byte{
		[]byte("dream"),
		[]byte("dreamer"),
		[]byte("erase"),
		[]byte("eraser"),
	}
	for i := range w {
		rev(w[i])
	}

	i := 0
	for i < len(sb) {
		var ok bool
		for _, t := range w {
			if bytes.HasPrefix(sb[i:], t) {
				ok = true
				i += len(t)
				break
			}
		}
		if !ok {
			fmt.Println("NO")
			return
		}
	}
	fmt.Println("YES")
}

func rev(b []byte) {
	for i, j := 0, len(b)-1; i < j; i, j = i+1, j-1 {
		b[i], b[j] = b[j], b[i]
	}
}
